qemu/tests/create_external_snapshot_via_driver.py

```diff
@@ -0,0 +1,135 @@
+        process.system(params.get("httpd_start_cmd"), shell=True)
+        out = process.system_output(params.get("httpd_status_cmd"), shell=True)
+        if not re.search(params.get("httpd_status_re"), out):
+            err = "Http server fails to start, because of {}".format(out)
+            exceptions.TestError(err)
+
+    subcommand = params.get("subcommand")
+    eval("%s_test()" % subcommand)
+
+    user_profile = ""
+    user_profile_backup = ""
+    if params.get("external_snapshot_protocol") == "ssh":
+        user_profile = params.get("user_profile")
+        user_profile_backup = params.get("user_profile_backup")
+        shutil.copy(user_profile, user_profile_backup)
+
+    base_image = params.get("images", "image1").split()[0]
+    params.update(
+        {"image_name_%s" % base_image: params["image_name"],
+         "image_format_%s" % base_image: params["image_format"]})
+    t_file = params["guest_file_name_%s" % base_image]
+    snapshot_test = qemu_disk_img_info.InfoTest(
+        test, params, env, base_image)
+    logging.info("STEP 2: Save file md5sum before create snapshot.")
+    snapshot_test.start_vm(params)
+    md5_before_snapshot = snapshot_test.save_file(t_file)
+    if not md5_before_snapshot:
+        raise exceptions.TestError("Fail to save tmp file.")
+    snapshot_test.destroy_vm()
+
+    logging.info("STEP 3: Start to create snapshot")
+    snapshot_test_2 = qemu_disk_img_info.InfoTest(test, params, env, "sn1")
+    n_params = snapshot_test_2.create_snapshot()
+    if params.get("external_snapshot_protocol") == "ssh":
+        ssh_cfg = process.system_output("cat /root/.bash_profile", shell=True)
+        ssh_sock = re.search("(SSH_AUTH_SOCK)=(.*?);", ssh_cfg)
+        ssh_pid = re.search("(SSH_AGENT_PID)=(\d+);", ssh_cfg)
+        command_prefix = "%s=%s;export %s;"
+        command_prefix += "%s=%s;export %s;"
+        ssh_auth_sock = ssh_sock.group(1)
+        ssh_sock_path = ssh_sock.group(2)
+        ssh_agent_pid = ssh_pid.group(1)
+        ssh_pid_num = ssh_pid.group(2)
+
+        n_params["qemu_command_prefix"] = command_prefix % (
+            ssh_auth_sock,
+            ssh_sock_path,
+            ssh_auth_sock,
+            ssh_agent_pid,
+            ssh_pid_num,
+            ssh_agent_pid)
+
+    remove_https_proxy = params.get("remove_https_proxy")
+    if params.get("external_snapshot_protocol") == "https":
+        n_params["qemu_command_prefix"] = "%s" % remove_https_proxy
+
+    logging.info("STEP 4: Start vm using external snapshot created")
+    snapshot_test.start_vm(n_params)
+    output = snapshot_test.info()
+    snapshot_test.check_backingfile(output)
+
+    logging.info("STEP 5: Compare md5 value")
+    md5_after_snapshot = snapshot_test.check_file(t_file, md5_before_snapshot)
+    if not md5_after_snapshot:
+        raise exceptions.TestError("Check md5sum fail (file:%s)" % t_file)
+    snapshot_test.destroy_vm()
+
+    if params.get("external_snapshot_protocol") == "ssh":
```

Comment:
Better to use try/finally to clean the environment. And shall we need to stop http service?